src/Libraries/Microsoft.Extensions.AI/ChatCompletion/ChatClientStructuredOutputExtensions.cs

```diff
@@ -224,4 +225,21 @@ private static JsonSerializerOptions GetOrCreateDefaultJsonSerializerOptions()
     [JsonSerializable(typeof(JsonNode))]
     [JsonSourceGenerationOptions(WriteIndented = true)]
     private sealed partial class JsonNodeContext : JsonSerializerContext;
+
+    /// <summary>
+    /// Remove characters from type name that are valid in metadata but shouldn't be used in a schema name.
+    /// Removes arrays and generic type parameters, and replaces invalid characters with underscores.
+    /// </summary>
+    private static string SanitizeMetadataName(string typeName) =>
+        InvalidNameCharsRegex().Replace(typeName, "_");
+
+    /// <summary>Regex that flags any character other than ASCII digits or letters or the underscore.</summary>
+#if NET
+    [GeneratedRegex("[^0-9A-Za-z_]")]
+    private static partial Regex InvalidNameCharsRegex();
+#else
+    private static Regex InvalidNameCharsRegex() => _invalidNameCharsRegex;
+    private static readonly Regex _invalidNameCharsRegex = new("[^0-9A-Za-z_]", RegexOptions.Compiled);
+#endif
```

Comment:
Thanks for the fix.

Could you move the equivalent logic that's currently in AIFunctionFactory to instead be on the internal FunctionCallHelpers class and then use that from both there and here (rather than duplicating the code)?  We may subsequently choose to expose some of the functionality publicly, too.

Reply:
done!